net/interfaces/interfaces.go

```diff
@@ -407,7 +425,7 @@ func GetState() (*State, error) {
 			return
 		}
 		for _, pfx := range pfxs {
-			if pfx.IP().IsLoopback() || pfx.IP().IsLinkLocalUnicast() {
+			if pfx.IP().IsLoopback() {
```

Comment:
This changes the behavior of the `Interfaces.HaveV4` field which is documented as:

```go
        // HaveV4 is whether the machine has some non-localhost,                                                                                                                            
        // non-link-local IPv4 address on a non-Tailscale interface that's up.                                                                                                              
        HaveV4 bool
```

That comment would no longer be true.

If this part needs to change, then docs should be updated and all callers should be audited to make sure they're cool with whatever the new semantics are.

There's a maxim I've heard "never reuse a name for a new meaning" that suggests "HaveV4" would be better renamed to something else if we give it a new meaning, so break any code written assuming the old semantics. That might be a little strict in this case, as this is mostly an internal detail and we can find all the callers (as opposed to a public library where we don't know who the callers are).

Reply:
HaveV4 and HaveV6 are only used in netcheck and decide whether to send IPv4 and/or IPv6 probes. That still seems like the appropriate use, we're now broadening the set of addresses which we'll probe.

Renamed HaveV6Global to just HaveV6, as ULA is acceptable for connectivity.

Reply:
So my new concern is `AnyInterfaceUp`, which is used by a lot of things to determine how we go into idle mode.

With this change, IPv4 linklocal addresses (as assigned by OSes when there's no internet/DHCP server) get link local IPv4, and then this new code sets HaveV4, which means AnyInterfaceUp is now true, and all the various things using it now think the world is up, when it's not:

```
$ git grep AnyInterfaceUp
health/health.go:// SetAnyInterfaceUp sets whether any network interface is up.
health/health.go:func SetAnyInterfaceUp(up bool) {
ipn/ipnlocal/local.go:  networkUp := ifst.AnyInterfaceUp()
ipn/ipnlocal/local.go:  networkUp := b.prevIfState.AnyInterfaceUp()
logtail/logtail.go:     return l.linkMonitor.InterfaceState().AnyInterfaceUp()
logtail/logtail.go:             if st.AnyInterfaceUp() {
net/interfaces/interfaces.go:// AnyInterfaceUp reports whether any interface seems like it has Internet access.
net/interfaces/interfaces.go:func (s *State) AnyInterfaceUp() bool {
net/interfaces/interfaces.go:   if s.AnyInterfaceUp() {
wgengine/userspace.go:  e.magicConn.SetNetworkUp(e.linkMon.InterfaceState().AnyInterfaceUp())
wgengine/userspace.go:  up := cur.AnyInterfaceUp()
wgengine/userspace.go:  health.SetAnyInterfaceUp(up)
```

I confirmed that this happens on Windows:

<img width="680" alt="Screen Shot 2021-06-08 at 8 41 32 AM" src="https://user-images.githubusercontent.com/2621/121216219-8d0a0c80-c835-11eb-8495-a023573c3eff.png">

Can we make this new behavior conditional on running in Cloud Run / Heroku / etc environments?

We could move the what-environment-are-we accessor somewhere else (controlclient was always kinda weird) and make it cheaper with a sync.Once or atomic.Value, like `hostinfo_windows.go`'s `func osVersionWindows`?

Reply:
It is avahi-autoipd that assigns link-local address if there is no regular address, and it doesn't set IFF_DYNAMIC. It *does* set the address to link scope.

```
2: eth0@if64: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 4a:77:06:d8:75:e5 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 169.254.10.89/16 brd 169.254.255.255 scope link eth0:avahi
       valid_lft forever preferred_lft forever
    inet6 fe80::4877:6ff:fed8:75e5/64 scope link 
       valid_lft forever preferred_lft forever
```

----
AWS Lambda configures a 169.254.x.x address but sets it to global scope.
```
9: vinternal_1@if10: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
  link/ether 7e:7b:f2:9e:c2:1d brd ff:ff:ff:ff:ff:ff link-netnsid 0
  inet 169.254.76.1/23 scope global vinternal_1
  valid_lft forever preferred_lft forever
```

----
I think we can try that:
- Linux only
- 169.254.x.x address with scope global.

Reply:
Let's try this:
- interfaces_linux.go adds a isIp4LinkLocalUsable() function to check if a link-local address has global scope.
- it is only called from LocalAddresses() which I think happens rarely, so it creates a fresh RTNETLINK socket each time. We expect most devices won't have any 169.254.x.x addresses, so it won't be called at all.
- add an interfaces_default.go for all other platforms which just returns false.

I've verified that an AWS Lambda node can get connectivity using this.